sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_compiler_opencl.asciidoc

```diff
@@ -0,0 +1,253 @@
+forms of `kernel_bundle::has_kernel` or `kernel_bundle::get_kernel` that take a
+type-name are not useful for kernels defined in OpenCL C.
+
+
+== Example
+
+The following example shows a simple SYCL program that defines an OpenCL C
+kernel as a string and then compiles and launches it.
+
+```
+#include <sycl/sycl.hpp>
+namespace syclex = sycl::ext::oneapi::experimental;
+
+int main() {
+  sycl::queue q;
+
+  // Kernel defined as an OpenCL C string.  This could be dynamically
+  // generated instead of a literal.
+  std::string source = R"""(
+    __kernel void my_kernel(__global int *in, __global int *out) {
+      size_t i = get_global_id(0);
+      out[i] = in[i]*2 + 100;
+    }
+  )""";
+
+  sycl::kernel_bundle<sycl::bundle_state::ext_oneapi_source> kb_src =
+    syclex::create_kernel_bundle_from_source(
+      q.get_context(),
+      syclex::source_language::opencl,
+      source);
+
+  // Compile and link the kernel from the source definition.
+  sycl::kernel_bundle<sycl::bundle_state::executable> kb_exe =
+    syclex::build(kb_src);
+
+  // Get a "kernel" object representing the kernel defined in the
+  // source string.
+  sycl::kernel k = kb_exe.ext_oneapi_get_kernel("my_kernel");
+
+  constexpr int N = 4;
+  int input[N] = {0, 1, 2, 3};
+  int output[N] = {};
+
+  sycl::buffer inputbuf(input, sycl::range{N});
+  sycl::buffer outputbuf(output, sycl::range{N});
+
+  q.submit([&](sycl::handler &cgh) {
+    sycl::accessor in{inputbuf, cgh, sycl::read_only};
+    sycl::accessor out{outputbuf, cgh, sycl::read_write};
+
+    // Each argument to the kernel is a SYCL accessor.
+    cgh.set_args(in, out);
+
+    // Invoke the kernel over a range.
+    cgh.parallel_for(sycl::range{N}, k);
+  });
+}
+```
+
+
+== Issues
+
+* How should we expose the difference between OpenCL C versions?
+  It seems like there are two aspects to the problem.
+  Applications need some way to query which versions the backend (or device)
+  supports.
+  Applications also need some way to tell the runtime which version the kernel
+  is written in.
++
+--
+One option is to define separate enumerators in `source_language` for each
+version like this:
+
+```
+enum class source_language : /*unspecified*/ {
+  opencl_1_0,
+  opencl_1_1,
+  opencl_2_0,
+  opencl_3_0,
+};
+```
+
+Applications could then query the supported versions via
+`is_source_kernel_bundle_supported`, and applications would identify the
+version of their kernel string via the `lang` parameter to
+`create_kernel_bundle_from_source`.
+
+Alternatively, this extension could define just a single language enumerator
+(`opencl`), but also provide as separate query to get the supported OpenCL C
+versions.
+When building a kernel bundle, applications would be required to pass "-cl-std"
+via the `build_options` property in order to identify the OpenCL C version of
+their source string.
+--
+
+* How can an application determine the OpenCL C optional features that are
```

Comment:
This actually can be quite important for some users. Per my understanding the kernel compiler is intended for solving the problem with compilation of OpenCL code when only Level Zero backend is available (now you always need OpenCL backend for that). Given the above, the ability to query information about OpenCL features could be quite important.

Reply:
We plan to add support to the kernel compiler for writing kernels in SYCL.  I'd rather move people towards this, rather than adding a bunch of OpenCL-specific queries.  If people write kernels in SYCL, they can use the existing SYCL queries to determine which features are supported.

Reply:
> I'd rather move people towards this

While I agree with the intention there are projects that have tens of kernels written in OpenCL for historical reasons. Rewriting those kernels would require a lot of resources that the projects may not have so we have to take the situations into account.

Reply:
We don't plan to drop support for OpenCL kernels.  I'm just saying that we shouldn't invest more resources into further improvements for these kernels.  The existing OpenCL kernel support doesn't provide any way to query about the optional features (beyond using OpenCL directly), and existing applications that use OpenCL kernels have apparently survived without this feature.

Reply:
Now, the applications require to have the OpenCL backend to work. In order to get rid of the requirement we need to be able to compile OpenCL kernels without OpenCL backend that will be achieved by introducing the kernel compiler. But this will solve only a half of the problem. The kernel compiler will not be fully usable for compiling the OpenCL kernels unless we can query the information about what we can actually compile with it.

Reply:
@gmlueck, now, that you are reworking the extension we (oneDNN) have a few requirements for it:
* The extension should be able to compile OpenCL C code when there is no OpenCL backend and only Level Zero backend is available
* We need to be able to query OpenCL extensions when there is no OpenCL backend and only Level Zero backend is available
* Compilation time with the extension should be on par with the compilation time with the OpenCL API (the current online compiler seems to meet the requirement)

If any of the above requirements is not met then the extension will be practically useless for oneDNN.

Reply:
We have been discussing this issue offline.  @densamoilov, I'd like to leave this issue unresolved in this PR and address it in a follow-on PR.  @cperkinsintel already has an implementation for the spec in its current state, so I'd like to get the current spec and implementation merged.  I'll add new APIs in a follow-on PR, and Chris can implement them in a follow-on implementation PR.

If that is OK, I'd like to resolve this comment.  Note that the spec already lists this as an unresolved issue, so that will be an additional reminder to address this.

Reply:
Sounds good, thank you. I'm resolving this.